core/src/main/java/org/jdbi/v3/core/rewriter/InternalRewrittenStatement.java

```diff
@@ -36,49 +37,55 @@
-                        .orElseThrow(() -> {
-                            String msg = String.format("Unable to execute, no named parameter matches " +
-                                            "\"%s\" and no positional param for place %d (which is %d in " +
-                                            "the JDBC 'start at 1' scheme) has been set.",
-                                    named_param, index, index + 1);
-                            return new UnableToExecuteStatementException(msg, context);
-                        });
+            bindNamed(params, statement);
+        }
+    }
+
+    private void bindPositional(Binding params, PreparedStatement statement) {
+        int paramCount = stmt.getParams().size();
+        for (int i = 0; i < paramCount; i++) {
+            int index = i;
+            Argument a = params.findForPosition(i)
+                    .orElseThrow(() -> {
+                        String msg = String.format("Unable to execute, no positional parameter bound at position %s",
+                                index);
+                        return new UnableToExecuteStatementException(msg, context);
+                    });
+            try {
+                a.apply(i + 1, statement, this.context);
+            } catch (SQLException e) {
+                throw new UnableToExecuteStatementException(
+                        String.format("Exception while binding positional param at (0 based) position %d",
+                                i), e, context);
+            }
+        }
+    }
+
+    private void bindNamed(Binding params, PreparedStatement statement) {
+        List<String> paramList = stmt.getParams();
+
+        if (paramList.contains(POSITIONAL_PARAM)) {
+            String msg = "Cannot mix named and positional parameters in a SQL statement: "
+                    + Joiner.on(", ").join(paramList);
+            throw new UnableToExecuteStatementException(msg, context);
+        }
+
+        for (int i = 0; i < paramList.size(); i++) {
+            String param = paramList.get(i);
+            Argument a = params.findForName(param)
+                    .orElseThrow(() -> {
+                        String msg = String.format("Unable to execute, no named parameter matches \"%s\".", param);
+                        return new UnableToExecuteStatementException(msg, context);
+                    });
 
-                try {
-                    a.apply(i + 1, statement, this.context);
-                } catch (SQLException e) {
-                    throw new UnableToCreateStatementException(String.format("Exception while binding '%s'",
-                            named_param), e, context);
-                }
-                i++;
+            try {
+                a.apply(i + 1, statement, this.context);
+            } catch (SQLException e) {
+                throw new UnableToCreateStatementException(String.format("Exception while binding '%s'",
```

Comment:
Nitpick: I think we need to decide with with quotes we wrap the parameter in error messages (single or double quotes).